Server: Split the huge CGameContext::OnMessage()

Kaffeine opened the pull request:
This is actually helpful for forks and modding: some mods customize votes, some customize the processing of `NETMSGTYPE_CL_SAY`, etc.

It is hard to track what is going on in this method, and sometimes it is even harder to merge changes here.

This MR is hard to rebase and I'd appreciate a quick (like: before other MRs touching this area) response. :eyes: 

<s>**Note: I can't fix the code style for this MR, this is why I asked for #7033 ("please update clang-format").**</s>

Edit: Oh, what!? It failed in my repo and I gave up. Now it goes :heavy_check_mark: here :exploding_head: 

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

def- reviewed (approved):
Seems fine for me, but this split will also cause lots of effort for forks, right?

Robyt3 commented on src/game/server/gamecontext.cpp:
```diff
@@ -1926,455 +1926,17 @@ void CGameContext::OnMessage(int MsgID, CUnpacker *pUnpacker, int ClientID)
+			OnSetTeamNetMessage(pRawMsg, ClientID);
 		else if(MsgID == NETMSGTYPE_CL_ISDDNETLEGACY)
 		{
 			IServer::CClientInfo Info;
```
For consistency, we should extract handling of every message into a separate function, even if that function ends up being a single line long.

Kaffeine replied:
I agree. I stopped here after a few hours of work to get feedback and to waste less time in case of decline.
The further split planned for further PRs, is it that necessary to have everything in one PR?

Robyt3 replied:
You can also do it in a separate PR, I just wanted to make sure that we agree on the final result of the refactoring.

Kaffeine replied:
:heavy_check_mark:

Robyt3 commented on src/game/server/gamecontext.cpp:
```diff
@@ -1926,455 +1926,17 @@ void CGameContext::OnMessage(int MsgID, CUnpacker *pUnpacker, int ClientID)
 	if(Server()->ClientIngame(ClientID))
 	{
+		CPlayer *pPlayer = m_apPlayers[ClientID];
 		if(MsgID == NETMSGTYPE_CL_SAY)
```
This could probably become a `switch`.

Kaffeine replied:
Sure (planned for the next PR). I can make this PR bigger if you really needed it but I think that in terms of MRs "split" is better than "squash".

heinrich5991 replied:
I think for me, if I had to resolve this merge conflict, it'd be easier to do if this was done all in one commit.

Kaffeine replied:
> it'd be easier to do if this was done all in one commit
> We're deciding to move everything out.

Having a net message per commit allows to solve conflicts per net message.
E.g. if a mod changed only the `SAY` handling then other commits 'll be applied smoothly.

In general:
- It is easier to solve conflicts in less diff, so smaller commits are better
- It is inconvenient to solve the same conflict again and again, so (in general) the changes in the same area should be in the same commit.

The rule of thumb is that the "subsequent commits should not change the content of the previous commits".

heinrich5991 replied:
My idea would be that it'd be in the original commits. Having it all in one commit would also make it easier to retroactively apply little changes to each of the commits.

Kaffeine replied:
> Having it all in one commit would also make it easier to retroactively apply little changes to each of the commits.

:100: agree

Robyt3 commented on src/game/server/gamecontext.cpp:
```diff
@@ -1926,455 +1926,17 @@ void CGameContext::OnMessage(int MsgID, CUnpacker *pUnpacker, int ClientID)
-				SendChat(ClientID, Team, aCensoredMessage, ClientID);
-			}
-		}
+			OnSayNetMessage(pRawMsg, pUnpacker, ClientID);
```
I'd prefer if we perform the downcast before calling the functions, so each individual `On...Message` function already gets the specific type that it handles as argument.

Kaffeine replied:
Meh. I see why we can want it but also it is disappointing to have `#include <game/generated/protocol.h>` in `gamecontext.h` because GameContext is included by virtually all entities and it makes the dependency tree much bigger and causes many more recompilation on protocol details changed.

I'd rather find a way to get rid of this https://github.com/ddnet/ddnet/commit/b99d8dc2595d0914b9a844dad7b98b0aaedca8a8#diff-9bc2378398eb6b7e1ab70a095b206eccef287f58209d704b1aff0650ea3cf426R18.

heinrich5991 replied:
You can either forward-declare these classes or just call the parameters `class CNetMsg_Cl_Say`, then you don't need the forward-declaration.

Kaffeine replied:
:heavy_check_mark:

heinrich5991 commented:
> This is actually helpful for forks and modding: some mods customize votes, some customize the processing of `NETMSGTYPE_CL_SAY`, etc.

Is it helpful for mods?

It's definitely a huge merge conflict for existing mods, so it should actually improve the situation if it should get merged. @ChillerDragon Do you have an opinion on this?

ChillerDragon commented:
Yea merging this into forks will be a pain. But seems like a nice cleanup also for the ddnet codebase directly.
If #7127 gets implemented and these new methods could be overwritten by a custom controller. All of this could be modded with zero conflict surface. That would have to be another pr with conflicts then I guess.

But all in all :+1: thanks for pushing that to the upstream @Kaffeine

Kaffeine commented:
I extracted the remaining methods and replaced the if/else with `switch()`.

Now with `switch` it is disappointing that `NETMSGTYPE` is not an enum class.

Kaffeine commented:
> Seems fine for me, but this split will also cause lots of effort for forks, right?

It depends.

The changes in logger cause a lot of pain. The changes in [CCharacterCore::m_HookedPlayer](https://github.com/ddnet/ddnet/commit/f07d773492f60ca22fda41854b425aec352b7537) caused some pain because was hard to find.
And e.g. those [1](https://github.com/ddnet/ddnet/commit/808f889c00a189d957202895a87597c41127d76e#diff-65908fcaf323f624c0a8e584e92664d0d49640bb139729c6e22a14a5b2170c3fR1734) (2018), [2](aede6df2ca3b86b5b8c197731f0ca9d3c63031f5) (old)) changes in timing were disaster because they didn't break the compilation but led to subtly weirdest bugs and odd CPU load. I spent two weeks investigating those issues when tried to move the codebase to DDNet.

The changes in CGameContext::OnMessage() are straightforward. They causes conflicts and won't cause hidden bugs. It is clear what the forks need to do here. The sync can take some time (ten minutes?) but I won't say it is "lots of effort or forks".

heinrich5991 commented on src/game/server/gamecontext.cpp:
```diff
@@ -1926,772 +1926,840 @@ void CGameContext::OnMessage(int MsgID, CUnpacker *pUnpacker, int ClientID)
+			break;
+		}
+	}
+	if(MsgID == NETMSGTYPE_CL_STARTINFO)
```
Should be part of the `switch`.

Kaffeine replied:
Ok, what to do with the `if(Server()->ClientIngame(ClientID))` check which is used for all net messages except this one?

Options:
- Add to each `case`
- Add to each `On$$NetMessage()`

heinrich5991 replied:
Oh, didn't see that. I guess it's fine like it's now.

heinrich5991 commented on src/game/server/gamecontext.h:
```diff
@@ -288,6 +288,19 @@ class CGameContext : public IGameServer
+	void OnSetSpectatorModeNetMessage(const CNetMsg_Cl_SetSpectatorMode *pMsg, int ClientID);
+	void OnChangeInfoNetMessage(const CNetMsg_Cl_ChangeInfo *pMsg, int ClientID);
+	void OnEmoticonNetMessage(const CNetMsg_Cl_Emoticon *pMsg, int ClientID);
+	void OnKillNetMessage(int ClientID);
```
I think for symmetry, this should also get the `CNetMsg_Cl_Kill` parameter.

Kaffeine replied:
There is still `OnIsDDNetLegacyNetMessage(CUnpacker *pUnpacker, int ClientID)` which breaks it.
Also unfortunately OnSayNetMessage needs `const CUnpacker *pUnpacker` to record the `chillerbot` message in TeeHistorian.

Robyt3 replied:
The class `CNetMsg_Cl_Kill` might be empty right now, but we or a mod could add members to this class in the future, so having the parameter already in place seems beneficial.

There also appears to be an empty class `CNetMsg_Cl_IsDDNetLegacy` that you could use for `OnIsDDNetLegacyNetMessage`.

Robyt3 replied:
I'd also say `const CUnpacker *pUnpacker` should be the last argument, so the first two arguments are always the message and client ID.

Kaffeine replied:
:heavy_check_mark:

heinrich5991 reviewed (commented):
Looks generally good to me.

Kaffeine commented:
Hi. I think all the CR issues are fixed. What I can do to have this merged?

heinrich5991 reviewed (approved):
LGTM, thanks.
